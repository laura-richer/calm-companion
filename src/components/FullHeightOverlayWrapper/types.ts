export type FullHeightOverlayWrapperProps = {
  children: string | JSX.Element | JSX.Element[];
  show: boolean;
  title: string;
  onClose: () => void;
};
