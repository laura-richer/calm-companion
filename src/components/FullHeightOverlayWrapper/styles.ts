import { StyleSheet } from 'react-native';
import { spacer } from '~styles/variables';

export default StyleSheet.create({
  fullHeightOverlayHeader: {
    marginTop: spacer.XS,
    marginBottom: spacer.MD,
  },
  backButton: {
    position: 'absolute',
    zIndex: 1,
  },
  fullHeightOverlayTitle: {
    maxWidth: '80%',
    textAlign: 'center',
    marginHorizontal: 'auto',
  },
});
