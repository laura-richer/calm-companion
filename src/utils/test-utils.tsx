import React, { PropsWithChildren } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { render, RenderOptions } from '@testing-library/react-native';
import { Provider as StoreProvider } from 'react-redux';
import { EnhancedStore, configureStore } from '@reduxjs/toolkit';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { RootState } from '~store/index';
import navigationReducer, {
  initialState as navigationInitialState,
} from '~store/slices/navigation';
import quickHelpReducer, { initialState as quickHelpInitialState } from '~store/slices/quickHelp';

type DeepPartial<T> = {
  [K in keyof T]?: T[K] extends object ? DeepPartial<T[K]> : T[K];
};

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: EnhancedStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  { preloadedState = {}, store = createTestStore(preloadedState) }: ExtendedRenderOptions = {}
) => {
  const wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return (
      <StoreProvider store={store}>
        <SafeAreaProvider>
          <NavigationContainer>{children}</NavigationContainer>
        </SafeAreaProvider>
      </StoreProvider>
    );
  };

  return {
    ...render(ui, {
      wrapper,
    }),
  };
};

const _createStore = (preloadedState: Partial<RootState> = {}): EnhancedStore => {
  return configureStore({
    reducer: {
      navigation: navigationReducer,
      quickHelp: quickHelpReducer,
    },
    preloadedState,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        immutableCheck: { warnAfter: 256 },
        serializableCheck: { warnAfter: 256 },
      }),
  });
};

export const createTestStore = (preloadedState: DeepPartial<RootState> = {}): EnhancedStore => {
  const state: RootState = {
    // @ts-ignore
    navigation: {
      ...navigationInitialState,
      ...(preloadedState.navigation || {}),
    },
    // @ts-ignore
    quickHelp: {
      ...quickHelpInitialState,
      ...(preloadedState.quickHelp || {}),
    },
  };
  const store: EnhancedStore = _createStore(state);
  const origDispatch = store.dispatch;
  store.dispatch = jest.fn(origDispatch);
  return store;
};
