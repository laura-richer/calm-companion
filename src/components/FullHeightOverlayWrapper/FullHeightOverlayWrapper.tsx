import React from 'react';
import { Modal, View, Text, Pressable } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import BackButton from '~components/BackButton/BackButton';
import { FullHeightOverlayWrapperProps } from './types';
import styles from './styles';
import { typographyStyles } from '~styles/global';

const FullHeightOverlayWrapper = ({
  children,
  show,
  onClose,
  title,
}: FullHeightOverlayWrapperProps) => {
  const insets = useSafeAreaInsets();

  return (
    <Modal animationType="slide" visible={show ? true : false} onRequestClose={onClose}>
      <View
        style={{
          paddingTop: insets.top,
          paddingBottom: insets.bottom,
          paddingLeft: insets.left,
          paddingRight: insets.left,
        }}
      >
        <View style={styles.fullHeightOverlayHeader}>
          <Pressable
            accessibilityLabel="Navigate back to previous screen"
            style={styles.backButton}
            onPress={() => onClose()}
          >
            <BackButton type="dark" />
          </Pressable>
          <Text style={[typographyStyles.textMd, styles.fullHeightOverlayTitle]}>{title}</Text>
        </View>
        <View>{children}</View>
      </View>
    </Modal>
  );
};

export default FullHeightOverlayWrapper;
