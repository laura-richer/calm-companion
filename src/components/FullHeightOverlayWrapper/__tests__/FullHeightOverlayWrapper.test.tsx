import React from 'react';
import { Text } from 'react-native';
import { screen, fireEvent } from '@testing-library/react-native';
import { renderWithProviders } from '~utils/test-utils';
import FullHeightOverlayWrapper from '../FullHeightOverlayWrapper';

const mockCloseOverlay = jest.fn();

const _renderDefault = () => {
  return renderWithProviders(
    <FullHeightOverlayWrapper show={true} onClose={mockCloseOverlay} title="Full height modal">
      <Text>Child content</Text>
    </FullHeightOverlayWrapper>
  );
};
describe('full height overlay', () => {
  beforeEach(() => {
    _renderDefault();
  });

  test('matches snapshot', () => {
    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('renders child content', () => {
    const content = screen.getByText('Child content');
    expect(content).toBeOnTheScreen();
  });

  test('show title', () => {
    const title = screen.getByText('Full height modal');
    expect(title).toBeOnTheScreen();
  });

  test('close modal on back button click', () => {
    const backButton = screen.getByLabelText('Navigate back to previous screen');
    expect(backButton).toBeOnTheScreen();

    fireEvent.press(backButton);
    expect(mockCloseOverlay).toHaveBeenCalled();
  });
});
