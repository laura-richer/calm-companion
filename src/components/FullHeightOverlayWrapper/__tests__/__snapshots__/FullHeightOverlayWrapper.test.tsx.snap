// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`full height overlay matches snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  visible={true}
>
  <View
    style={
      {
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
      }
    }
  >
    <View
      style={
        {
          "marginBottom": 30,
          "marginTop": 10,
        }
      }
    >
      <View
        accessibilityLabel="Navigate back to previous screen"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "position": "absolute",
            "zIndex": 1,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            {
              "testUri": "../../../src/assets/images/back-button-dark.png",
            }
          }
          style={
            {
              "height": 25,
              "width": 30,
            }
          }
        />
      </View>
      <Text
        style={
          [
            {
              "fontFamily": "Poppins-Bold",
              "fontSize": 14,
            },
            {
              "marginHorizontal": "auto",
              "maxWidth": "80%",
              "textAlign": "center",
            },
          ]
        }
      >
        Full height modal
      </Text>
    </View>
    <View>
      <Text>
        Child content
      </Text>
    </View>
  </View>
</Modal>
`;
